### PR TITLE
You might also like ...

### PR DESCRIPTION
**Pull request by manno**
Adds JSON metadata to event so we can store:

* remote id (from frab i.e.)
* related events

Adds the related events to the postroll html page.  Related videos links can be generated via the halfnarp importer.
The remote id is filled from schedule XML and if present removed from the tags.

### CODE REVIEW

**Comment by MaZderMind** [comment]
nice! I'll try to review & merge today and make the post-page pretty in the next week :) Nice feature!

**Comment by manno** [comment]
I'll have to run two commands manually to take this live:

    UpdateRemoteId.new.perform(Event.all.pluck(:id))
    load 'lib/halfnarp.rb'

Works for me 😃

File: lib/halfnarp.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  next if weights.count < 2
+  limit = percentile(weights, 0.95)
+  edges.delete_if { |_k, v| v < limit.to_i }
+}
+
+def find_event(conference, id)
+  conference.events.find_by('metadata @> ?', { remote_id: id.to_s }.to_json)
+end
+
+conference = Conference.find_by(acronym: '33c3')
```
Comment (MaZderMind) [inline lib/halfnarp.rb]: You shouldn't hardcode a single conference-name here. Instead there should be a task that updates all halfnarp backed conferences. Whether a conference supports that and where the relevant halfnarp is located should be a property of the conference.

File: app/models/concerns/fahrplan_updater.rb [hunk 2]
```diff
@@ -3,20 +3,26 @@ module FahrplanUpdater
...
+    self.metadata[:remote_id] = id
+    self.link = get_event_url(conference, id)
+    update_attributes info
+  end
+
+  private
+
+  def get_event_url(conference, id)
+    return unless conference.schedule_url.present?
+    conference.schedule_url.sub('schedule.xml', "events/#{id}.html").freeze
```
Comment (MaZderMind) [inline app/models/concerns/fahrplan_updater.rb]: Why did you move this getter fro mthe conference-model, as it seems to require the conference anyway (now as an argument)

Also, this logic will only work for schedules.xmls directly generated by a frab. Often we build our own schedule.xml with scripts on a c3voc-host, where this assumption does not hold, but as you didn't change this in this PR this might be the wrong place to change that.

File: lib/halfnarp.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
```
Comment (MaZderMind) [inline lib/halfnarp.rb]: Shouldn't this be a sidekiq-task which updates the metadata when a new conferences / release is added?

File: lib/update_remote_id.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Update events remote_id from existing schedule xml data
```
Comment (MaZderMind) [inline lib/update_remote_id.rb]: Shouldn't this be a sidekiq-task which updates the remote-id when a new conferences / release is added?

File: lib/halfnarp.rb (hunk 3 above)
Comment (manno) [inline lib/halfnarp.rb]: I think this is only needed once for 33c3. The majority of conferences will not have halfnarp data.

File: lib/halfnarp.rb (hunk 1 above)
Comment (manno) [inline lib/halfnarp.rb]: Might turn this into a rake task for 34c3 to load halfnarp votes from a local file.

File: lib/halfnarp.rb (hunk 3 above)
Comment (MaZderMind) [inline lib/halfnarp.rb]: well we have halfnarp-data for 32c3 and we will have it for 34c3. Also it is not good practice to hardcode event-names into a software, that is usable and might be used by anyone (isn't this the reason we renamed it from media.ccc.de to voctoweb?)

File: lib/update_remote_id.rb (hunk 4 above)
Comment (manno) [inline lib/update_remote_id.rb]: I'll delete this after I've updated the live database. The modified importer will set the remote id if schedule.xml data is available.

File: app/models/concerns/fahrplan_updater.rb (hunk 2 above)
Comment (manno) [inline app/models/concerns/fahrplan_updater.rb]: Also, I'm not sure why you build a schedule.xml at all. Can't you set all required fields via the API?

**Comment by manno** [comment]
merged 6331795

File: lib/halfnarp.rb (hunk 3 above)
Comment (MaZderMind) [inline lib/halfnarp.rb]: I personally don't have the 32c3 data but I guessed it should still be around.
Well why not call `rake halfnarp:import 33c3 https://…/votes.json`

This way at least other people (like me) could import data for another event later, without changing the code on the production system or in the repo.

File: lib/update_remote_id.rb (hunk 4 above)
Comment (MaZderMind) [inline lib/update_remote_id.rb]: But if you delete this file, how should this mapping work for later conferences? Why not fix it the right way now and not have to think about it again next year?

No one of us will know how to run these tasks again next year and no one will maintain them…

**Comment by MaZderMind** [comment]
:( why did ask for feedback then anyway?

File: lib/halfnarp.rb (hunk 3 above)
Comment (manno) [inline lib/halfnarp.rb]: If you want that, please open another issue. We can discuss it there.

For starters:
* There is no URL for halfnarp data.
* The halfnarp file I received by mail needs to be edited to be valid JSON.
* `halfnarp.rb` is not even a class and not meant to be reused.
* Dropping everything not above the 95 percentile, might result in unexpected data for smaller conferences.
* Only linking to related events within the same conference might not be what we want long term.

File: lib/halfnarp.rb (hunk 1 above)
Comment (manno) [inline lib/halfnarp.rb]: You're supposed to `load` this in the rails console. There is no argv available. 

It's actually a common practice to maintain data on websites this way, I can move it to 'db/data/' to make this more explicit. The alternative is me not committing this at all to the public repo.

File: lib/update_remote_id.rb (hunk 4 above)
Comment (manno) [inline lib/update_remote_id.rb]: This is called from the API controller:

https://github.com/voc/voctoweb/blob/63317953452fcdddbcfc36bddb40ef4db2f5df5a/app/models/concerns/fahrplan_updater.rb#L17

However I should add the metadata to https://github.com/voc/voctoweb/blob/431ed8efd86d6403fa8aa99c078438b2d383e8f6/app/controllers/api/events_controller.rb#L93 so you can set it directly via the API.
